pandemic51/core/streaming.py

```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
+        to disk in `pandemic51.config.STREAMS_DIR`.
+
+        Args:
+            stream_name: the stream name
+
+        Returns:
+            the chunk path
+        '''
+        self.chunk_path = _get_chunk_url(self.webpage)
+        self.write_json(self.path, pretty_print=True)
+
+    def _attempt_get_uris(self):
+        return m3u8.load(self.chunk_path).segments.uri
+
+    @classmethod
+    def _from_dict(cls, d):
+        stream_name = d["stream_name"]
+        GMT = d["GMT"]
+        webpage = d["webpage"]
+        chunk_path = d["chunk_path"]
+        return cls(stream_name, GMT, webpage, chunk_path)
+
+
+class MjpegStream(Stream):
+    '''A Stream class that reads MJPEGs'''
+    def __init__(self, stream_name, GMT, url):
+        super(MjpegStream, self).__init__(stream_name, GMT)
+        self.url = url
+
+    def get_live_stream_url(self):
+        '''Get the URL for streaming'''
+        return self.url
+
+    def download_image(self, outdir):
+        '''Downloads an image from the latest stream
+
+        Args:
+            outdir: the output directory
+
+        Returns:
+            is_new_img: `True` if the image was not already on disk
+            image_path: path the the downloaded image on disk
+            dt: datetime object of when the image was downloaded
+        '''
+        dt = datetime.utcnow()
+
+        # UTC integer timestamp (epoch time)
+        timestamp = int(dt.timestamp())
+
+        # Create path for image
+        image_path = os.path.join(
+            outdir, self.stream_name, "%d.jpg" % timestamp)
+        etau.ensure_basedir(image_path)
+
+        # Capture the current frame of the stream
+        is_new_img = sample_first_frame(self.url, image_path)
+
+        return is_new_img, image_path, dt
+
+    @classmethod
+    def _from_dict(cls, d):
+        stream_name = d["stream_name"]
+        GMT = d["GMT"]
+        url = d["url"]
+        return cls(stream_name, GMT, url)
+
+
+class ImageStream(Stream):
+    '''A Stream class for streams that intermittently take image snapshots,
+    rather than providing constant video feed.
     '''
-    with etau.TempDir(basedir=panc.BASE_DIR) as tmpdir:
-        # Download video
-        video_path, dt = download_chunk(stream_name, tmpdir)
+
+    def __init__(self, stream_name, GMT, webpage, url_filter):
+        super(ImageStream, self).__init__(stream_name, GMT)
+        self.webpage = webpage
+        self.url_filter = url_filter
+
+    def get_live_stream_url(self):
+        '''Get the URL for streaming'''
+        # Get the source from the page
+        urls = get_img_urls(self.webpage)
+        filtered_urls = [u for u in urls if self.url_filter in u]
+
+        if not filtered_urls:
+            raise Exception("No URLs found for webpage: %s" % self.webpage)
+
+        # we only need one!
+        url = filtered_urls[0]
+
+        # Get the large version of the image instead of the thumbnail
```

Comment:
This is really only applicable to the Ypsilanti livestream, the rest of this class seems pretty general though

Reply:
I'm going to leave it until we add another `ImageStream` if you're okay with that